### PR TITLE
Vue checks 🤠

### PR DESCRIPTION
**Pull request by moralesbang**
## Summary

- Added first part of vue checks with some react checks fixes.
- Added `"no-extra-parens": "off"` config for eslint to avoid eslint formatting conflicts.

## Screenshots

Screen Shot 2020-05-13 at 6 26 56 PM

## Trello Card

https://trello.com/c/q11PdEEr

### CODE REVIEW

File: src/frontEnd/vueChecks/createSummary.js [hunk 1]
```diff
@@ -1,5 +1,205 @@
...
+    metric: 'SUMMARY-BUILDING-5',
+    description: 'El proyecto genera archivos minificados en el build',
+    value: 'N/A'
+  });
+
+  summary.push({
+    metric: 'SUMMARY-BUILDING-6',
+    description:
+      'El proyecto usa webpack para generar el build en producción y babel para los imports con alias',
+    value: reports.find(({ metric }) => metric === vueMetrics.USE_CLI_SERVICE).value
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/createSummary.js:116]: `reports.some(({ metric, value }) => metric === vueMetrics.USE_CLI_SERVICE && value)`

File: src/frontEnd/vueChecks/createSummary.js [hunk 2]
```diff
@@ -1,5 +1,205 @@
...
+      reports.foEarch(elem => {
+        if (elem.metric === dependenciesMetrics.VUEX && elem.value) {
+          hasVuexDependency = true;
+        }
+        if (elem.metric === vueMetrics.USE_VUEX) {
+          hasVuexImport = true;
+        }
+      });
+
+      return hasVuexDependency && hasVuexImport;
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/createSummary.js:192]: You can do this:
```
reports.some(elem.metric === dependenciesMetrics.VUEX && elem.value) && 
reports.some(elem.metric === dependenciesMetrics. USE_VUEX && elem.value)
```

File: src/frontEnd/vueChecks/utils.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+};
+
+module.exports.checkVueTemplate = async testPath => {
+  const badFiles = await findSync(/[Vv]ue.component\(/, `${testPath}/src`, '.js$');
+  return !Object.keys(badFiles).length;
+};
+
+module.exports.checkVuexUse = async testPath => {
+  const response = await find(/^import [Vv]uex from 'vuex'/, `${testPath}/src/store`, 'index.js$')
+    .then(results => console.log('Vuex results', results) || !!Object.keys(results).length)
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/utils.js]: Why are you adding the console log?

File: src/frontEnd/vueChecks/utils.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    .catch(() => false);
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/utils.js:27]: I think this is unnecesary

File: src/frontEnd/vueChecks/utils.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  return Object.values(results).every(({ line }) => line[0].includes('scoped'));
+};
+
+module.exports.checkVueTemplate = async testPath => {
+  const badFiles = await findSync(/[Vv]ue.component\(/, `${testPath}/src`, '.js$');
+  return !Object.keys(badFiles).length;
+};
+
+module.exports.checkVuexUse = async testPath => {
+  const response = await find(/^import [Vv]uex from 'vuex'/, `${testPath}/src/store`, 'index.js$')
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/utils.js]: Why are you using find and not `findSync`?

File: src/generalChecks/constants.js [hunk 6]
```diff
@@ -17,14 +17,26 @@ module.exports.generalMetrics = createObject([
...
 
 module.exports.folderStructure = {
   react: ['app', 'config', 'constants', 'redux', 'scss', 'services', 'utils'],
-  vue: ['components', 'config', 'constants', 'scss', 'services', 'store', 'utils', 'views'],
+  vue: [
+    'assets',
+    'components',
+    'config',
+    'constants',
+    'mixins',
```
Comment (LucasZibell) [inline src/generalChecks/constants.js:33]: I would remove this. It is not always neccesary

File: src/generalChecks/constants.js [hunk 7]
```diff
@@ -17,14 +17,26 @@ module.exports.generalMetrics = createObject([
...
+  'I18N'
 ]);
 
 module.exports.DOCKERFILE_ATTRIBUTES = ['FROM', 'WORKDIR', 'COPY', 'RUN', 'ENV', 'WORKDIR'];
 
 module.exports.folderStructure = {
   react: ['app', 'config', 'constants', 'redux', 'scss', 'services', 'utils'],
-  vue: ['components', 'config', 'constants', 'scss', 'services', 'store', 'utils', 'views'],
+  vue: [
+    'assets',
```
Comment (LucasZibell) [inline src/generalChecks/constants.js:29]: I would remove this. It is not always neccesary

File: src/generalChecks/constants.js [hunk 8]
```diff
@@ -17,14 +17,26 @@ module.exports.generalMetrics = createObject([
   'CODE_OWNERS',
   'FOLDER_STRUCTURE',
   'BABEL_IMPORTS',
-  'CODE_QUALITY'
+  'CODE_QUALITY',
+  'I18N'
```
Comment (LucasZibell) [inline src/generalChecks/constants.js:21]: Why did you add this here?

File: src/frontEnd/reactChecks/createSummary.js [hunk 9]
```diff
@@ -6,7 +6,6 @@ const dependenciesMetrics = require('../../dependenciesChecks/constants');
 const seoMetrics = require('../seoChecks/constants');
 
 const { generalMetrics } = require('../../generalChecks/constants');
-const dependenciesMetrics = require('../../dependenciesChecks/constants');
```
Comment (LucasZibell) [inline src/frontEnd/reactChecks/createSummary.js]: Why are you removing this?? it is used in the first metrics

File: src/frontEnd/vueChecks/index.js [hunk 10]
```diff
@@ -28,5 +29,13 @@ module.exports = async testPath => {
     value: await checkVuexUse(testPath)
   });
 
+  vueResult.push({
+    metric: vueMetrics.STATE_MODULES,
+    description: 'El proyecto usa modulos para separar el estado de la aplicación',
+    value: fs.existsSync(`${testPath}/src/store/modules`)
+  });
+
+  console.log('✨VUE CHECKS ✨\n\n', vueResult);
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/index.js:38]: Remove the console.log

File: src/generalChecks/constants.js (hunk 6 above)
Comment (moralesbang) [reply src/generalChecks/constants.js:33]: Ok

File: src/generalChecks/constants.js (hunk 7 above)
Comment (moralesbang) [reply src/generalChecks/constants.js:29]: Ok

File: src/frontEnd/reactChecks/createSummary.js [hunk 11]
```diff
@@ -93,6 +94,13 @@ const buildingSummary = (summary, reports) => {
     description: 'Se utiliza el package de deploy para la gestión de releases',
     value: 'Manual'
   });
+
+  summary.push({
+    metric: 'SUMMARY-BUILDING-4',
+    description: 'Se corre el checkeo de linter tanto antes de generar un build como de push',
+    value: reports.some(({ metric, value }) => metric === eslintMetrics.ESLINT_ERRORS && !!value)
```
[the remark this answers is not in the document]
Comment (moralesbang) [reply src/frontEnd/reactChecks/createSummary.js:100]: Yeah of course, good look.

File: src/generalChecks/constants.js (hunk 8 above)
Comment (moralesbang) [reply src/generalChecks/constants.js:21]: Unexpected issue

File: src/frontEnd/vueChecks/utils.js (hunk 4 above)
Comment (moralesbang) [reply src/frontEnd/vueChecks/utils.js:17]: I put that for the case that `src/store` doesn't exists.

File: src/frontEnd/vueChecks/createSummary.js (hunk 2 above)
Comment (moralesbang) [reply src/frontEnd/vueChecks/createSummary.js:193]: Your suggestion is so clear and valid, but not performant, because you're doing 2 loops for just one validation that it's possible to do with just one. Do you still wanna change that?

File: src/frontEnd/vueChecks/utils.js (hunk 5 above)
Comment (moralesbang) [reply src/frontEnd/vueChecks/utils.js]: I've done that in aims to use `then/catch`.

File: src/frontEnd/vueChecks/createSummary.js [hunk 12]
```diff
@@ -1,5 +1,212 @@
...
+const testMetrics = require('../../testChecks/constants');
+const envMetrics = require('../../envChecks/constants');
+const { generalMetrics } = require('../../generalChecks/constants');
+const { vueMetrics } = require('../vueChecks/constants');
+const { eslintMetrics } = require('../../linterChecks/constants');
+const dependenciesMetrics = require('../../dependenciesChecks/constants');
+const seoMetrics = require('../seoChecks/constants');
+
+const limits = {
+  reduxRecomposePercentage: 70,
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/createSummary.js:11]: I think you are not using this

File: src/frontEnd/vueChecks/createSummary.js [hunk 13]
```diff
@@ -1,5 +1,212 @@
...
+    description:
+      'La arquitectura de la aplicación se encuentra preparada para implementar test de integración',
+    value: 'N/A'
+  });
+
+  summary.push({
+    metric: 'SUMMARY-TESTING-3',
+    description: 'Hay una cobertura del 70% o más en los test unitarios',
+    value: reports.some(
+      elem => elem.metric === testMetrics.CODE_COVERAGE && elem.value >= limits.minTestCoverage
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/createSummary.js:33]: I think this doesnt exist `limits.minTestCoverage `

File: src/frontEnd/vueChecks/createSummary.js [hunk 14]
```diff
@@ -1,5 +1,212 @@
...
+      elem => elem.metric === generalMetrics.OUTDATED_DEPENDENCIES && elem.value.includes('vue')
+    )
+  });
+
+  summary.push({
+    metric: 'SUMMARY-BUILDING-2',
+    description: 'Las dependencias del proyecto están actualizadas',
+    value:
+      reports.filter(elem => elem.metric === generalMetrics.UNUSED_DEPENDENCIES).length <=
+      limits.maxUnusedDependencies
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/createSummary.js:92]: I think this doesnt exist limits. maxUnusedDependencies

File: src/frontEnd/vueChecks/createSummary.js [hunk 15]
```diff
@@ -1,5 +1,212 @@
...
+    value: !reports.some(
+      elem => elem.metric === generalMetrics.OUTDATED_DEPENDENCIES && elem.value.includes('vue')
+    )
+  });
+
+  summary.push({
+    metric: 'SUMMARY-BUILDING-2',
+    description: 'Las dependencias del proyecto están actualizadas',
+    value:
+      reports.filter(elem => elem.metric === generalMetrics.UNUSED_DEPENDENCIES).length <=
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/createSummary.js:91]: Why are you using `filter`? Shouldnt it be `some`??

File: src/frontEnd/vueChecks/createSummary.js [hunk 16]
```diff
@@ -1,5 +1,212 @@
...
+    metric: 'SUMMARY-BUILDING-5',
+    description: 'El proyecto genera archivos minificados en el build',
+    value: 'N/A'
+  });
+
+  summary.push({
+    metric: 'SUMMARY-BUILDING-6',
+    description:
+      'El proyecto usa webpack para generar el build en producción y babel para los imports con alias',
+    value: reports.some(({ metric, value }) => metric === vueMetrics.USE_CLI_SERVICE && value)
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/createSummary.js:117]: I think this is not the correct check 🤔

File: src/frontEnd/vueChecks/createSummary.js [hunk 17]
```diff
@@ -1,5 +1,212 @@
...
+  });
+};
+
+const uiUxSummary = (summary, reports) => {
+  summary.push({
+    metric: 'SUMMARY-UI-UX-1',
+    description: 'El proyecto es mobile friendly según lighthouse',
+    value: reports.some(
+      elem => elem.metric === seoMetrics.LIGHTHOUSE_PWA_OVERALL && elem.value >= limits.pwaMin
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/createSummary.js:126]: This doesn't exist `limits.pwaMin `

File: src/frontEnd/vueChecks/createSummary.js [hunk 18]
```diff
@@ -1,5 +1,212 @@
...
+  summary.push({
+    metric: 'SUMMARY-UI-UX-4',
+    description: 'Proyecto respeta la estructura de directorios sugerida',
+    value: reports.filter(elem => elem.metric === generalMetrics.FOLDER_STRUCTURE).every(elem => elem.value)
+  });
+
+  summary.push({
+    metric: 'SUMMARY-UI-UX-5',
+    description: 'El proyecto usa vue-loader para generar componentes de un solo archivo SFC',
+    value: reports.some(elem => elem.metric === vueMetrics.USE_VUEX && elem.value)
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/createSummary.js:151]: Im pretty sure this is not the correct check 🤔

File: src/frontEnd/vueChecks/createSummary.js [hunk 19]
```diff
@@ -1,5 +1,212 @@
...
+  });
+
+  summary.push({
+    metric: 'SUMMARY-UI-UX-7',
+    description: 'SFC que no estén al nivel de app deben estar scoped',
```
Comment (LucasZibell) [inline src/frontEnd/vueChecks/createSummary.js:156]: Rewrite this description. All the SFC should be scoped

File: src/frontEnd/vueChecks/createSummary.js (hunk 2 above)
Comment (LucasZibell) [reply src/frontEnd/vueChecks/createSummary.js:193]: Okey noted, in that case i think that what you want to use is a `reduce`

File: src/utils/index.js [hunk 20]
```diff
@@ -8,8 +8,9 @@ exports.createObject = keyArray => keyArray.reduce((acc, key) => ({ ...acc, [key
 
 exports.analyzeMatches = matches => Object.keys(matches).filter(key => removeComments(matches[key]).length);
 
-exports.calculatePercentage = (results, total) =>
-  // eslint-disable-next-line prettier/prettier
-  this.analyzeMatches(results).length / total * percentage;
+exports.calculatePercentage = (results, total, options = { skipComments: false }) => {
```
Comment (LucasZibell) [inline src/utils/index.js:11]: Are you sure about this? `options = { skipComments: false }` I don't think it will work well as a default for other objects 🤔

File: src/frontEnd/vueChecks/createSummary.js (hunk 15 above)
Comment (moralesbang) [reply src/frontEnd/vueChecks/createSummary.js:91]: I've doing the same like react's `createSummary`. The answer is because the value of `UNUSED_DEPENDENCIES`'s report its the list of packages unused. Let's check this code snippet from `dependenciesCheck`:

```
  packages.forEach(dependency => {
    const { moduleName, latest, packageJson, unused, bump } = dependency;
    if (unused) {
      dependenciesResults.push({
        metric: dependenciesMetrics.UNUSED_DEPENDENCIES,
        description: 'Dependencia no usada',
        value: moduleName
      });
    } else if (bump && bump !== 'patch') {
      dependenciesResults.push({
        metric: dependenciesMetrics.OUTDATED_DEPENDENCIES,
        description: 'Dependencia no actualizada',
        value: `${moduleName} Version: packageJson: ${packageJson} -> ultima ${latest}`
      });
    }
  });
```

File: src/frontEnd/vueChecks/createSummary.js (hunk 14 above)
Comment (moralesbang) [reply src/frontEnd/vueChecks/createSummary.js:92]: I'm using the value in `react/createSummary`.

File: src/frontEnd/vueChecks/createSummary.js (hunk 19 above)
Comment (moralesbang) [reply src/frontEnd/vueChecks/createSummary.js:156]: Cool

File: src/frontEnd/vueChecks/createSummary.js (hunk 2 above)
Comment (moralesbang) [reply src/frontEnd/vueChecks/createSummary.js:193]: I'm gonna use:

```
summary.push({
  metric: 'SUMMARY-CLIENT-SERVER-4',
  description: 'Usa Vuex store para manejar el estado de la aplicación ',
  value: reports.some(({ metric, value }) => metric === vueMetrics.USE_VUEX && value)
});
```
Because if the project use Vuex inherently it have it as depency.

File: src/utils/index.js (hunk 20 above)
Comment (moralesbang) [reply src/utils/index.js:11]: I'm gonna change it, for a simple approach.
